src/main/java/com/icthh/xm/ms/scheduler/manager/SchedulingManager.java

```diff
@@ -212,6 +218,25 @@ private boolean deleteTaskFromTenant(String tenant, String taskKey,
         return cancelled;
     }
 
+    /**
+     * Set state for task.
+     * Method init tenant context with value from input task because method called from another thread
+     *
+     * @param state new state of the task
+     * @param task  the task changing
+     */
+    public void setState(StateKey state, TaskDTO task) {
+        Long taskId = task.getId();
+        log.info("Marking task as finished id: {}, tenant: {}", task.getId(), task.getTenant());
+        TenantContextUtils.setTenant(tenantContextHolder, task.getTenant());
```

Comment:
I think there we can get unexpected behavior with setting tenant. Because setTenant() just set its value to thread, and i believe these threads are in the pool so could be reused. Thus in multi tenant environment this thread will have the Tenant value initialized from previous run.
I propose to use Privileged context execute method instead:
```
tenantContextHolder.getPrivilegedContext().execute(tenant, {} -> {...})
```

Reply:
ok, i will create another pull request

Reply:
https://github.com/xm-online/xm-ms-scheduler/pull/11